kyaml/fn/framework/matchers.go

```diff
@@ -0,0 +1,342 @@
+	return &TemplatedMetaSliceMatcher{
+		Templates: names,
+		MetaMatcher: func(names sets.String, meta yaml.ResourceMeta) bool {
+			return names.Has(meta.Name)
+		},
+	}
+}
+
+// NamespaceMatcher matches resources whose metadata.namespace are equal to one of the provided values.
+// e.g. `NamespaceMatcher("foo", "bar")` matches if `metadata.namespace` is either "foo" or "bar".
+//
+// NamespaceMatcher supports templating.
+// e.g. `NamespaceMatcher("{{.AppName}}")` will match `metadata.namespace` "foo" if TemplateData is
+// `struct{ AppName string }{ AppName: "foo" }`
+func NamespaceMatcher(names ...string) ResourceTemplateMatcher {
+	return &TemplatedMetaSliceMatcher{
+		Templates: names,
+		MetaMatcher: func(names sets.String, meta yaml.ResourceMeta) bool {
+			return names.Has(meta.Namespace)
+		},
+	}
+}
+
+// KindMatcher matches resources whose kind is equal to one of the provided values.
+// e.g. `KindMatcher("foo", "bar")` matches if `kind` is either "foo" or "bar".
+//
+// KindMatcher supports templating.
+// e.g. `KindMatcher("{{.TargetKind}}")` will match `kind` "foo" if TemplateData is
+// `struct{ TargetKind string }{ TargetKind: "foo" }`
+func KindMatcher(names ...string) ResourceTemplateMatcher {
+	return &TemplatedMetaSliceMatcher{
+		Templates: names,
+		MetaMatcher: func(names sets.String, meta yaml.ResourceMeta) bool {
+			return names.Has(meta.Kind)
+		},
+	}
+}
+
+// APIVersionMatcher matches resources whose kind is equal to one of the provided values.
+// e.g. `APIVersionMatcher("foo/v1", "bar/v1")` matches if `apiVersion` is either "foo/v1" or
+// "bar/v1".
+//
+// APIVersionMatcher supports templating.
+// e.g. `APIVersionMatcher("{{.TargetAPI}}")` will match `apiVersion` "foo/v1" if TemplateData is
+// `struct{ TargetAPI string }{ TargetAPI: "foo/v1" }`
+func APIVersionMatcher(names ...string) ResourceTemplateMatcher {
+	return &TemplatedMetaSliceMatcher{
+		Templates: names,
+		MetaMatcher: func(names sets.String, meta yaml.ResourceMeta) bool {
+			return names.Has(meta.APIVersion)
+		},
+	}
+}
+
+// GVKMatcher matches resources whose API group, version and kind match one of the provided values.
+// e.g. `GVKMatcher("foo/v1/Widget", "bar/v1/App")` matches if `apiVersion` concatenated with `kind`
+// is either "foo/v1/Widget" or "bar/v1/App".
+//
+// GVKMatcher supports templating.
+// e.g. `GVKMatcher("{{.TargetAPI}}")` will match "foo/v1/Widget" if TemplateData is
+// `struct{ TargetAPI string }{ TargetAPI: "foo/v1/Widget" }`
+func GVKMatcher(names ...string) ResourceTemplateMatcher {
+	return &TemplatedMetaSliceMatcher{
+		Templates: names,
+		MetaMatcher: func(names sets.String, meta yaml.ResourceMeta) bool {
+			gvk := strings.Join([]string{meta.APIVersion, meta.Kind}, "/")
+			return names.Has(gvk)
+		},
+	}
+}
+
+// TemplatedMetaSliceMatcher is a utility type for constructing matchers that compare resource
+// metadata to a slice of (possibly templated) strings.
+type TemplatedMetaSliceMatcher struct {
```

Comment:
If the templating is used, right? Or is there another problem? Also, this is really a utility type out of which all the others are built. Is it worth saying on _all_ the types that use this?

Reply:
I don't remember why I wrote this, but I think it was because `m.values` gets modified whenever `InitTemplates` is called, which could cause a data race.

I think it's fine, no need to change anything.